### ISSUE TITLE
Organization locations assignment

### ISSUE THREAD

**Issue by githengi** [OpenSRP/opensrp-server-web#541]
Support the ability to do team assignment at the highest level, but look at the plans locations to determine which operational areas fall under that jurisdiction and assign the user.

Example plan seems the user was assigned to the country
https://web.reveal-stage.smartregister.org/plans/update/9f7f43f0-3155-5579-8b39-b93b34fcfad7

@moshthepitt @p-netm I think this assignment was done from reveal web,

There are 3 options for the assignment
1. Assign the user to all jurisdictions as they on the plan
2. Assign the user to the lowest common parent of the  jurisdictions in the plan
3. Using a hybrid of 1 and 2 where if the common parent in 2 is several levels up (maybe past particular levels) then have an algorithm return several parents instead of 1

For 2 (what I think is what web is using ) the solution should be to assign the  lowest common parent for the selected jurisdictions and not the root of the location hierarchy

Since there is already an API for the location hierarchy for a plan could this be used by the web or do you need additional API to help the web when doing the assignments.

**Comment by moshthepitt** [comment]
if we have a tree like this 

```
A -> B -> C -> E,F,G

A -> B -> Q -> P,L,K
```

What we want is that if in he UI the user selects A, what we should send to the API is B (the lowest common parent).  Right?

**Comment by githengi** [comment]
Yes exactly

**Comment by p-netm** [comment]
1 vs 2.
on the web: 

*Assign the user to all jurisdictions as they are on the plan*
- (walk tree to | get) find the node where assignment was done.
- walk whole `plan location hierarchy` starting at node that assignment was done; goal is to look for leaf nodes for assignment
- more expensive than 1

*Lowest Common Parent* 
- can do a breadth first search starting node where assignment was done on `plan location hierarchy`; goal is to look for the first jurisdiction that has more than one children locations or no children. 


3. a variation of 2 but with an extra requirement; a heuristic that dictates how to select the parents. this heuristic can only serve to make 2 more expensive and possible in the worst case 2 would perform just as worse or even more worse than 1 

@githengi @moshthepitt .

**Comment by moshthepitt** [comment]
@p-netm based on the above what is your recommendation?

**Comment by craigappl** [comment]
In the worst case scenario, we may be identifying 60k operational areas or more.

**Comment by githengi** [comment]
the problems with 1 @craigappl is someone wants to remove assignments they have to edit a lot of records

**Comment by githengi** [comment]
@p-netm I think 2 using BFS would work and would not be expensive. What is your thought on this?

**Comment by p-netm** [comment]
> @p-netm I think 2 using BFS would work and would not be expensive. What is your thought on this?

yeah 2 would be less expensive, it would unequivocally make assignment easier, at the moment am thinking about un-assignments which would have to push us to 3. 

This is how I envision it in my head, @githengi

![IMG20200922100636](https://user-images.githubusercontent.com/28119869/93853042-cb699680-fcbb-11ea-91dd-9ad1eb9518a4.jpg)

one question I have is in the 4th action step, would the already existing assignment to D have to revoked, or it can remain>

**Comment by p-netm** [comment]
> if we have a tree like this
> 
> ```
> A -> B -> C -> E,F,G
> 
> A -> B -> Q -> P,L,K
> ```
> 
> What we want is that if in he UI the user selects A, what we should send to the API is B (the lowest common parent). Right?

@moshthepitt should the user assign as above, would this mean that the user will not see the assignment on A but has to drilldown to B, wouldn't this seem unexpected, and probably look like a bug.

**Comment by moshthepitt** [comment]
@p-netm yeah that is a good point.  I am considering allowing assignments ONLY for leaf nodes.  That paired with the ability to select all should work. No?

**Comment by p-netm** [comment]
@moshthepitt by 

> select all

do you mean like create a payload of all leaf nodes once, the user assigns at the root level, or just send a single payload of the jurisdiction as rootJurisdiction.id>

**Comment by moshthepitt** [comment]
no I meant hide the form for everything apart from leaf nodes

**Comment by p-netm** [comment]
cool.

**Comment by p-netm** [comment]
I need help figuring out which of this options has a more reasonable compomise,
the goal is to be able to perform team assignments at higher levels up the tree such that the assignments would not behave in a buggy way as described  [in this slack thread](https://onaio.slack.com/archives/CBSU8L2SC/p1600399197005900).

- the first option: once assignment occurs on a non-leaf node; the web will collate all the OAs that both belong to the node and those within the plan hierarchy and create a payload of those.
 
the compromise here include:
- potential performance hit for a sufficiently large hierarchy.
- unassignments will have to take place at each of the operational areas.

the alternative which is the LCP(lowest Common Parent) has the compromise that:
- during un-assignments: we could have some performance hits depending on the size and depth of the tree.
- during assignments: an inconsistency could be observed where the assignments will not be shown on the node that the assignment happened as described [here](https://github.com/OpenSRP/opensrp-server-web/issues/541#issuecomment-696675132).

in terms of performance: the second has a slight edge over the first.

**Comment by ukanga** [comment]
Could the un-assignment be done on OpenSRP API instead? You just need to tell OpenSRP to remove assignment for this user for this location and the decision up or down the tree is made by OpenSRP server?

**Comment by githengi** [comment]
that is possible, however it should be remove the location and all its descendants i.e downwards

**Comment by ukanga** [comment]
Great, to reduce the heavy lifting from the front end, I would recommend we address that decision from the backend. What would that API look like? Would this change your concerns @p-netm ?

**Comment by p-netm** [comment]
>     * during assignments: an inconsistency could be observed where the assignments will not be shown on the node that the assignment happened as described [here](https://github.com/OpenSRP/opensrp-server-web/issues/541#issuecomment-696675132).

@ukanga @githengi this is a concern for the FE due to the LCP assignments approach.  previously, the assignment we sent, targeted the node where assingment happened. i.e if the user assigned at the root node, we would send an assignment payload reflecting that regardless of whether the root node was the Lowest common parent of all OAs assigned to the plan.

thus when we later get the payload we can truthfully reflect the assignment on the web. ie. we display it where the user expects it to be, at the level that they assigned it.

with the the LCP approach there is a chance that the assignment payload we send and later receives targets a node that is lower in the tree than where the assignment happened(i.e the node that the user assigned the team to). <- this is the compromise that I need to know if its ok.

**Comment by githengi** [comment]
@p-netm I think in that case it, the web should send 2 API calls, the 1st un-assign the current assignment.  The 2nd the new assignment at the lower level.

**Comment by craigappl** [comment]
Peter, Sam and I discussed this on a call.

Decision: We need to deal with this on the OpenSRP server instead of the web UI. It may be valuable to store the assignment at the plan itself. We should have a different set of tables that show the effective assignment. If you're assigning at the Zambia level, but the plan has one operational area, you need to store that you were assigned at Zambia, but the security/authenticate endpoint needs to read from a different table.

Short term:
- For each assignment that the user has, we would get the plan, query the locations of the plan, and identify all of the children.

Long term:

Workflow:
- Assign at Zambia level
- Receive the Zambia level and plan Id
- OpenSRP server identifies lowest common parent
- OpenSRP adds to a table called "Effective Assignment" for that organization
- The Security Authenticate endpoint reads from the effective Assignment table

Work to be done:
1. We need to modify the OpenSRP server to be able to collect two points. First, the assignment at whatever level is assigned. Second, the actual OA assignment.

Risks:
- If you edit the location hierarchy, this will break everything.
- If you deactivate or move locations, you would need to 

Additional thought:
- Enhancement - In the web UI, we could present a tree, allowing them to choose which level they want using a checkbox and we would do an assignment.

**Comment by githengi** [comment]
The solution/target for now is for the the security/authenticate and /user-assignment APIs to return plan locations within user organizations assigned jurisdictions and plans.

The solution will involve
- Query the user assigned plans and jurisdictions as currently is
- Get the user assigned hierarchy for assigned jurisdictions (set A)
- Get the plan locations hierarchy for assigned plans (set B). New query being introduced
- Return an intersection of the two sets

**Comment by craigappl** [comment]
Test steps:
1. Login as onatest user - Currently assigned to plan name Dynamic-IRS-2020-09-22-Permissions and mln_6 OA only
2. (Failed) Add user assignment to kmb_5 and sync your android client to make sure it shows up.
3. Remove both assignments and assign at the Zambia level in the web UI
4. Clear the app's data and login again
5. (Failed) View the app
6. Unassign from Zambia and assign at the Luapula Province
7. (Success) Clear the app data and the app
8. Add the assignment for Copperbelt Province so both provinces are set
9. (Failed - Error Processing Location) Clear the app data and view the app
10. Remove both provincial assignments and reassign at the Rural Health Centre level for Zambia>Copperbelt>Ndola
11. (Success) Clear the app data and view the app

Errors:
1. In step 2, the new location assignment in a different hierarchy didn't get updated in the android client.
2. In step 5, I got a Log In Failed
Here's the security Authenticate: http://reveal-stage.smartregister.org/opensrp/security/authenticate
```
{"message":"User not mapped on any location. Make sure that user is assigned to an organization with valid Location(s) ","status":"403 FORBIDDEN","data":null,"success":false}
```
Here's a screenshot: 
![image](https://user-images.githubusercontent.com/6067053/94758112-256bfb00-0351-11eb-86eb-0dfa0194491c.png)

Here's the assignedPlansAndLocations for this org https://reveal-stage.smartregister.org/opensrp/rest/organization/assignedLocationsAndPlans/0f38856a-6e0f-5e31-bf3c-a2ad8a53210d
```
[{"organizationId":"0f38856a-6e0f-5e31-bf3c-a2ad8a53210d","jurisdictionId":"22bc44dd-752d-4c20-8761-617361b4f1e7","planId":"932a8e5d-97bc-5d39-904e-7f9e63691bcf","fromDate":1601503200000,"toDate":1602453600000}]
```
Jurisdiction Info: https://reveal-stage.smartregister.org/opensrp/rest/location/22bc44dd-752d-4c20-8761-617361b4f1e7?is_jurisdiction=true&return_geometry=false
```
{"type":"Feature","id":"22bc44dd-752d-4c20-8761-617361b4f1e7","properties":{"code":"22bc44dd-752d-4c20-8761-617361b4f1e7","status":"Active","name":"Zambia","geographicLevel":0,"version":0,"ISO_A3":"ZMB","ADMIN":"Zambia"},"serverVersion":1595923406402,"locationTags":[{"id":8,"name":"Country"}]}
```
3. Step 9 above failed because I couldn't select the location. I logged out and logged in again and still got the message "Error Processing Locations Hierarchy". Below is the security authenticate:
```
{"jurisdictionIds":["10bb0c79-e5b5-4115-8836-c02e4d7514bc"],"locations":{"locationsHierarchy":{"map":{"22bc44dd-752d-4c20-8761-617361b4f1e7":{"id":"22bc44dd-752d-4c20-8761-617361b4f1e7","label":"Zambia","node":{"locationId":"22bc44dd-752d-4c20-8761-617361b4f1e7","name":"Zambia","tags":["Country"],"attributes":{"geographicLevel":0},"voided":false},"children":{"e4762f01-3b4f-46e3-ae1c-1db8965b7b43":{"id":"e4762f01-3b4f-46e3-ae1c-1db8965b7b43","label":"Copperbelt","node":{"locationId":"e4762f01-3b4f-46e3-ae1c-1db8965b7b43","name":"Copperbelt","parentLocation":{"locationId":"22bc44dd-752d-4c20-8761-617361b4f1e7","voided":false},"tags":["Province"],"attributes":{"geographicLevel":1},"voided":false},"children":{"9e4f65dd-9210-4d6b-a4ca-4da9dae076a0":{"id":"9e4f65dd-9210-4d6b-a4ca-4da9dae076a0","label":"Ndola","node":{"locationId":"9e4f65dd-9210-4d6b-a4ca-4da9dae076a0","name":"Ndola","parentLocation":{"locationId":"e4762f01-3b4f-46e3-ae1c-1db8965b7b43","voided":false},"tags":["District"],"attributes":{"geographicLevel":2},"voided":false},"children":{"7d6a39dd-9b1c-4cee-8866-5caf9d0e3fe6":{"id":"7d6a39dd-9b1c-4cee-8866-5caf9d0e3fe6","label":"Masala New","node":{"locationId":"7d6a39dd-9b1c-4cee-8866-5caf9d0e3fe6","name":"Masala New","parentLocation":{"locationId":"9e4f65dd-9210-4d6b-a4ca-4da9dae076a0","voided":false},"tags":["Rural Health Centre"],"attributes":{"geographicLevel":3},"voided":false},"children":{"10bb0c79-e5b5-4115-8836-c02e4d7514bc":{"id":"10bb0c79-e5b5-4115-8836-c02e4d7514bc","label":"mln_6","node":{"locationId":"10bb0c79-e5b5-4115-8836-c02e4d7514bc","name":"mln_6","parentLocation":{"locationId":"7d6a39dd-9b1c-4cee-8866-5caf9d0e3fe6","voided":false},"tags":["Operational Area"],"attributes":{"geographicLevel":4},"voided":false},"parent":"7d6a39dd-9b1c-4cee-8866-5caf9d0e3fe6"}},"parent":"9e4f65dd-9210-4d6b-a4ca-4da9dae076a0"}},"parent":"e4762f01-3b4f-46e3-ae1c-1db8965b7b43"}},"parent":"22bc44dd-752d-4c20-8761-617361b4f1e7"}}}},"parentChildren":{"22bc44dd-752d-4c20-8761-617361b4f1e7":["e4762f01-3b4f-46e3-ae1c-1db8965b7b43"],"7d6a39dd-9b1c-4cee-8866-5caf9d0e3fe6":["10bb0c79-e5b5-4115-8836-c02e4d7514bc"],"e4762f01-3b4f-46e3-ae1c-1db8965b7b43":["9e4f65dd-9210-4d6b-a4ca-4da9dae076a0"],"9e4f65dd-9210-4d6b-a4ca-4da9dae076a0":["7d6a39dd-9b1c-4cee-8866-5caf9d0e3fe6"]}}},"team":{"identifier":"00dee865-35ba-5c38-81fc-314104f114e2","locations":[{"display":"mln_6","name":"mln_6","uuid":"5611c432-fb82-49f1-be1c-1db8965b7b43"}],"team":{"teamName":"Ona testOrg","organizationIds":[15.0],"display":"Ona testOrg","location":{"display":"mln_6","name":"mln_6","uuid":"5611c432-fb82-49f1-be1c-1db8965b7b43"},"uuid":"0f38856a-6e0f-5e31-bf3c-a2ad8a53210d"},"uuid":"867d2e5b-2c31-44b5-b4ca-a134a733694d"},"time":{"time":"2020-10-01 04:29:45","timeZone":"Africa/Harare"},"user":{"username":"onatest","roles":["ROLE_OPENMRS","ROLE_offline_access","ROLE_uma_authorization"],"permissions":["ROLE_OPENMRS","ROLE_offline_access","ROLE_uma_authorization"],"preferredName":"Ona Test User","baseEntityId":"867d2e5b-2c31-44b5-b4ca-a134a733694d","attributes":{},"voided":false},"jurisdictions":["mln_6"]}
```
It looks like only 1 province is being returned. In this case, Copperbelt is being returned, but Luapula used to be returned. Here's the assignedLocationsAndPlans
```
[{"organizationId":"0f38856a-6e0f-5e31-bf3c-a2ad8a53210d","jurisdictionId":"5611c432-fb82-49f1-be1c-1db8965b7b43","planId":"932a8e5d-97bc-5d39-904e-7f9e63691bcf","fromDate":1601503200000,"toDate":1602453600000},{"organizationId":"0f38856a-6e0f-5e31-bf3c-a2ad8a53210d","jurisdictionId":"e4762f01-3b4f-46e3-ae1c-1db8965b7b43","planId":"932a8e5d-97bc-5d39-904e-7f9e63691bcf","fromDate":1601503200000,"toDate":1602453600000}]
```

**Comment by githengi** [comment]
@craigappl from the steps at 5. There was no assignment for the user so the error returned should be as expected.
